gsy_framework/data_classes.py

```diff
@@ -434,50 +378,43 @@ def csv_values(self) -> Tuple:
-            if key_in_dict_and_not_none(trade_dict, "buyer"):
-                trade_dict["buyer"] = trade_dict["buyer"].serializable_dict()
-        return json.dumps(trade_dict, default=json_datetime_serializer)
+        return json.dumps(self.serializable_dict())
 
     @classmethod
     def from_json(cls, trade_string) -> "Trade":
         """De-serialize trade from json string."""
-        trade_dict = json.loads(trade_string)
-        if trade_dict["match_details"]["offer"] is not None:
-            trade_dict["offer"] = BaseBidOffer.from_json(trade_dict["match_details"]["offer"])
-        if trade_dict["match_details"]["bid"] is not None:
-            trade_dict["bid"] = BaseBidOffer.from_json(trade_dict["match_details"]["bid"])
-        trade_dict.pop("match_details")
-        if trade_dict.get("residual"):
-            trade_dict["residual"] = BaseBidOffer.from_json(trade_dict["residual"])
-        if trade_dict.get("creation_time"):
-            trade_dict["creation_time"] = str_to_pendulum_datetime(trade_dict["creation_time"])
-        else:
-            trade_dict["creation_time"] = None
-        if trade_dict.get("time_slot"):
-            trade_dict["time_slot"] = str_to_pendulum_datetime(trade_dict["time_slot"])
-        if trade_dict.get("offer_bid_trade_info"):
-            trade_dict["offer_bid_trade_info"] = (
-                TradeBidOfferInfo.from_json(trade_dict["offer_bid_trade_info"]))
+        return cls.from_dict(json.loads(trade_string))
+
+    @classmethod
+    def from_dict(cls, trade_dict: Dict) -> "Trade":
+        """
+        Create Trade object from a dict that contains nested unserializable structs
+        (including offer, bid, residual and datetimes).
+        """
+        trade_dict["offer"] = (
+            BaseBidOffer.from_json(trade_dict["offer"]) if trade_dict.get("offer") else None)
+        trade_dict["bid"] = (
+            BaseBidOffer.from_json(trade_dict["bid"]) if trade_dict.get("bid") else None)
+        trade_dict["residual"] = (
+            BaseBidOffer.from_json(trade_dict["residual"]) if trade_dict.get("residual") else None)
+
+        trade_dict["creation_time"] = (
+            str_to_pendulum_datetime(trade_dict["creation_time"])
+            if trade_dict.get("creation_time") else None)
+
+        trade_dict["time_slot"] = (
+            str_to_pendulum_datetime(trade_dict["time_slot"])
+            if trade_dict.get("time_slot") else None)
+
+        trade_dict["offer_bid_trade_info"] = (
+            TradeBidOfferInfo.from_json(trade_dict["offer_bid_trade_info"])
+            if trade_dict.get("offer_bid_trade_info") else None)
+
         if trade_dict["seller"] is not None:
             trade_dict["seller"] = TraderDetails.from_json(trade_dict["seller"])
         if trade_dict["buyer"] is not None:
             trade_dict["buyer"] = TraderDetails.from_json(trade_dict["buyer"])
-        return Trade(**trade_dict)
+        return cls.from_serializable_dict(trade_dict)
```

Comment:
The difference here is that `from_dict` can accept a dict that is not immediately serializable, but a nested dict with other serializable dataclasses. If you notice, the `from_serialziable_dict` assumes that the members "offer", "bid", "buyer" ... are instantiated objects, while the `from_dict` assumes that these are dicts and tries to serialize them before. I would not like to touch this more and merge these 2 though, because both are used in gsy-e and I would not like to change their interface again.

Reply:
Thanks for the clarification. All good.